### ISSUE TITLE
Focusing an isFocusable, yet not isTabbable element results in focusing the first element in the container

### ISSUE THREAD

**Issue by skriems** [focus-trap/focus-trap#611]
Hi there!

We are using skip links in our modal which leverages the focus-trap. It seems that the way #565 was fixed broke our skip link implementation :)

It basically looks like this:
```
<hr id="skip-link-target-top"  />         # visibility: hidden
<a href="#skip-link-target-bottom" />
...
<hr id="skip-link-target-bottom" />       # visibility: hidden
<a href="#skip-link-target-top" />
```

When the user clicks (tabs I should say - it's actually hidden as well) on the skip link, the corresponding target becomes visible and gets focused.

I created a codesandbox to showcase the behaviour: https://codesandbox.io/s/focus-trap-skip-links-eyz0g
It works with version 6.7.1 but not with version 6.7.2.

Basically https://github.com/focus-trap/focus-trap/blob/1700055ed7629aecbd5b8fb73cd93dc0f248cffc/index.js#L414 leads to `lastOfGroupIndex` to be set to `containerIndex` which a few lines later on sets the `destinationNode` to `destinationGroup.firstTabbableNode`.

I was wondering if we should try a different approach to improve our UX for both sighted and blind users, but on the otherhand, the underlying assumption which is made on line 418ff doesn't seem right to me? :thinking: From a userperspective I would expect the tabflow to continue from the focused element.

What do you think?

Best regards!

**Comment by skriems** [comment]
Btw, it should be possible to write tests for #565 and this one with [cypress-real-events](https://github.com/dmtrKovalenko/cypress-real-events). At least we have a working cypress test for our skip links that I could contribute.

**Comment by stefcameron** [comment]
@skriems Hey, thanks for reaching out, and thank you for the CodeSandbox, very helpful for context.

My first observation, in playing with the sandbox, is that because the HR is just before the Continue button, you just happened to not experience #565, because the HR was not the last element in the container.

I forked your sandbox https://codesandbox.io/s/focus-trap-skip-links-forked-8s33z and put the "skip to bottom" HR _after_ the Continue button, and set the version of focus-trap to v6.7.1. If you press Enter on the "skip to bottom" link now, and try to press Tab, you'll see you're stuck there. Shift+Tab works, and that was identified in #565.

Assuming your sandbox is representative enough of your actual scenario, then the bug was there, but your scenario just happened to avoid it.

That said, I see how v6.7.2 breaks your flow. I think this is because the fix I made in #588 is making an assumption that (while the target is _not_ found to be the last tabbable node in its group), since it's focusable but not tabbable, it should be considered the last tabbable node. And then we mess with the focus order.

So today, here, we have

```javascript
      ...

      const containerIndex = findIndex(state.tabbableGroups, ({ container }) =>
        container.contains(target)
      );

      ...

        if (
          lastOfGroupIndex < 0 &&
          (state.tabbableGroups[containerIndex].container === target ||
            (isFocusable(target) && !isTabbable(target)))
        )
```

and I'm thinking we now need this:

```javascript
      ...

      let containerIndex;
      const containerGroup = find(state.tabbableGroups, ({ container }, idx) => {
        if (container.contains(target)) {
          containerIndex = idx;
          return true;
        }
        return false;
      });

      ...

        if (
          lastOfGroupIndex < 0 &&
          (state.tabbableGroups[containerIndex].container === target ||
            (isFocusable(target) && !isTabbable(target) && containerGroup.lastTabbableNode === target))
        )
```

But there's a problem: These two statements are opposite: `!isTabbable(target) && containerGroup.lastTabbableNode === target` Because target isn't tabbable, there's no way it'll ever be identified as the `lastTabbableNode`.

So how do we _reliably_ figure out that the `target` is the "last node" in the container, visually, in the same way the browser would determine it to be for tab order. We can't just rely on DOM order because I think CSS can be used to affect how the browser interprets/determines tab order (and trying to mimic this in focus-trap would be losing proposition, because it'll be different across browsers, and may change at any point in time).

What do you think? Do you agree with the problem?

**Comment by stefcameron** [comment]
> Btw, it should be possible to write tests for #565 and this one with [cypress-real-events](https://github.com/dmtrKovalenko/cypress-real-events). At least we have a working cypress test for our skip links that I could contribute.

Thanks for the pointer. I'm aware of cypress-real-events. The problem with that library is its dependence/reliance upon the Chrome DevTools. That means tests written with it could only run on Chrome. And I'm not sure if they could run headless in the cloud (i.e. not sure if DevTools are available in Chrome headless), since we run our Cypress tests in GitHub Actions for CI.

**Comment by skriems** [comment]
Hey Stefan. Weird.. when I move the HR below the _continue_ button I escape out of the trap and land on the activate button. Anyhow I get your point regarding #565. We don't wanne get stuck there.

It's definately quite problematic and not trivial. I actually thaught about the same approach you tried and as you said the HR is not _tabbable_, hence not in the array of tabbable elements. So I wonder: Is the information of _tabbable_ elements enough to make a decision about whether a particular element is the last element in the focus-trap? 🤔 I'm not so sure about this. I could image a bunch of use cases in which an element is not _tabbable_ but got _focused_ in some way and hitting `Tab` on it, would result in this behaviour.

In fact, in its' initial "state" the HR is neither _visible_, nor _focusable_ and especially not _tabbable_  (omg 😂 ). But for some reason we make that happen. Technically the browser is not really interested in the fact, that the HR is not tabbable. It fires the onKeyDown event (target becomes `document.activeElement` (?) and focuses the next element in the tab order. So at the end of the day we need a way to determine if `event.target` is the last element in the trap. ... do we?

To me it seems, that I also need the information about focussable elements in the array. Then I could code something like

1. `event.target` is my last element in the array
2. `event.target` is not the last element, but it has no adjacent _tabbable_(!) elements

I guess (2) could become costly if _alot_ of only _focusable_ adjacent elements are in the trap. But thats probably an edge case. On the other hand if I know the index of the `event.target` at this point (not being the last element in the array) and the index of the last _tabbable_ element I could solve #565 and this one in the same sense. And it wouldn't be costly.

So when we hit Tab on some element `updateTabbableNodes` inside `checkTab` could also consider _focusable_ elements and do the above? 🤔 This should be possible, because at this point `event.target` and `document.activeElement` point to the HR with `tabindex=-1` and `visibility: visible`.

What do you think? Am I missing something?

**Comment by stefcameron** [comment]
> Hey Stefan. Weird.. when I move the HR below the _continue_ button I escape out of the trap and land on the activate button.

I did see that as well. Uncaught focus escape! But v6.7.2 also fixes that (inadvertently). And since the HR is last, then pressing Tab from it, though it executes the code from my fix, focus ends-up going where you'd expect it to (back to the top). But I realize your implementation wants to make it easy for the user to skip all the stuff in between and get to the Continue button.

Incidentally, why not just set focus to the tabbable Continue button? Isn't that where the user is wanting to go when they "skip to bottom" anyway?

---

Whenever we call `updateTabbableNodes()`, we could not call just `tabbable()` but could also call `focusable()`, which would give us a list of all focusable nodes, not just tabbable+focusable.

The problem remains, though, that all we have to go on here is DOM order, and because of CSS, I don't think that necessarily visual tab order. That's the crux of the problem. I can't think of a reliable way to mimic browser tab behavior, across all the browsers we aim to support, in a way that continues to match what browsers do. I admit, I don't know how they determine it.

As much as possible, focus-trap tries to get out of the way and let the browser decide where focus should go when the Tab key is pressed. Things get hairy, though, when we get to the edges, which is where focus-trap comes-in to coax focus to the right place in order to give the user the illusion that focus is effectively "trapped" inside one or more areas of the UI (e.g. a modal dialog).

And in this particular case, we're dealing with an edge that requires knowing more than we know.

**Comment by stefcameron** [comment]
@skriems Thanks for that. I did some more reading, and I guess I was blowing smoke, so to speak. 😅 DOM order is pretty much tab order -- but there might be one problem: positive tabindex. I haven't had a chance to play with this yet.

Still, please checkout #616 and LMK if it works for you so far.

**Comment by stefcameron** [comment]
@all-contributors add @skriems for bug
